### ISSUE TITLE
minor race in exists check

### ISSUE THREAD

**Issue by joeyh** [warner/magic-wormhole#105]
When receiving to a file, wormhole checks if the file exists and refuses to overwrite it. This is a useful check, and kind of a security guard. However, there's a race condition in this check: The file may not exist when _decide_destname checks it, but could be created immediately afterwards. Then, wormhole would overwrite the newly created file with the file it receives.

This race seems unlikely to be usefully exploitable, but it is a chink in the defenses and such minor holes can sometimes prove unexpectedly fatal. (Perhaps if there was some way for the attacker to delay the wormhole receive so it does not open the file until a carefully chosen time...)

I suggest opening the file with O_CREATE | O_EXCL which will fail atomically if it already exists. Or, since it seems to rename received files into place from a temp file, renameat2() with RENAME_NOREPLACE may be useful.

**Comment by warner** [comment]
Good catch.. I agree we should fix this if we can. I'm not sure Python exposes the necessary APIs, though.

For the initial temporary file, we could use `open(tmpfile, mode="x")`, to ask python for exclusive creation, except that only works on py3.3 or later. I think `f=os.fdopen(os.open(tmpfile, os.O_CREAT|os.O_EXCL|os.O_WRONLY), "wb")` would be the py2.7-compatible way to express that, and the python docs say that `O_EXCL` is available on windows too, so we probably don't need any per-OS special-casing.

For the atomic rename, though, I don't see any APIs that offer failure-upon-target-exists. The existing ones all seem aimed to permit atomic replacement of an existing (or not existing) file, and go to great lengths to overcome e.g. windows' reluctance to clobber an existing file, the exact opposite of what we want here.

I'm guessing that in C, on Linux, we'd want to do something like:

```
int dirfd = diropen(TARGETDIR);
int rc = renameat2(dirfd, "target.tmp", dirfd, "target", RENAME_NOREPLACE)
```

I'm not finding any Python stdlib bindings to `renameat2()`.. do you know of any? I see `os.replace()` (added in py3.3, not available in py2.7), which has a signature that comes close, but lacks the `flags=` option, which makes me think it binds to the old `renameat()` instead of `renameat2()`.
